File: esim-store/includes/verification.php
<?php
declare(strict_types=1);

function verification_generate_code(): string
{
    return (string) random_int(100000, 999999);
}

/** Generates + stores a code and sends it via Zadarma SMS. Falls back to a
 *  visible "demo code" when Zadarma isn't configured or the send fails, so
 *  checkout always stays usable while credentials aren't set yet. */
function verification_send(PDO $pdo, array $config, string $phone): array
{
    $code = verification_generate_code();
    $expires = (new DateTimeImmutable('+5 minutes'))->format(DateTimeInterface::ATOM);

    $pdo->prepare('DELETE FROM verification_codes WHERE phone = ?')->execute([$phone]);
    $pdo->prepare('INSERT INTO verification_codes (phone, code, expires_at, attempts) VALUES (?, ?, ?, 0)')
        ->execute([$phone, $code, $expires]);

    $configured = !empty($config['zadarma_key']) && !empty($config['zadarma_secret']);
    if (!$configured) {
        return ['ok' => true, 'demo' => true, 'demo_code' => $code];
    }

    try {
        zadarma_send_sms($config, $phone, "Код подтверждения {$config['site_name']}: $code");
        return ['ok' => true, 'demo' => false];
    } catch (Throwable $e) {
        error_log('[zadarma] sms send failed: ' . $e->getMessage());
        return ['ok' => true, 'demo' => true, 'demo_code' => $code, 'warning' => 'sms_send_failed'];
    }
}

function verification_check(PDO $pdo, string $phone, string $code): bool
{
    $stmt = $pdo->prepare('SELECT * FROM verification_codes WHERE phone = ?');
    $stmt->execute([$phone]);
    $row = $stmt->fetch();
    if (!$row) {
        return false;
    }
    if (new DateTimeImmutable($row['expires_at']) < new DateTimeImmutable('now')) {
        $pdo->prepare('DELETE FROM verification_codes WHERE phone = ?')->execute([$phone]);
        return false;
    }
    if ((int) $row['attempts'] >= 5) {
        return false;
    }

    $pdo->prepare('UPDATE verification_codes SET attempts = attempts + 1 WHERE phone = ?')->execute([$phone]);

    if (!hash_equals((string) $row['code'], $code)) {
        return false;
    }

    $pdo->prepare('DELETE FROM verification_codes WHERE phone = ?')->execute([$phone]);
    return true;
}

File: esim-store-django/store/verification.py
import logging
import secrets

from django.conf import settings
from django.utils import timezone

from . import zadarma
from .models import VerificationCode

logger = logging.getLogger(__name__)


def generate_code() -> str:
    return f"{secrets.randbelow(1_000_000):06d}"


def send(phone: str) -> dict:
    """Generates + stores a code and sends it via Zadarma SMS. Falls back to a
    visible "demo code" when Zadarma isn't configured or the send fails, so
    checkout always stays usable while credentials aren't set yet."""
    code = generate_code()
    expires_at = timezone.now() + timezone.timedelta(minutes=5)

    VerificationCode.objects.filter(phone=phone).delete()
    VerificationCode.objects.create(phone=phone, code=code, expires_at=expires_at, attempts=0)

    if not zadarma.is_configured():
        return {"ok": True, "demo": True, "demo_code": code}

    try:
        zadarma.send_sms(phone, f"Код подтверждения {settings.SITE_NAME}: {code}")
        return {"ok": True, "demo": False}
    except Exception:
        logger.exception("[zadarma] sms send failed")
        return {"ok": True, "demo": True, "demo_code": code, "warning": "sms_send_failed"}


def check(phone: str, code: str) -> bool:
    try:
        row = VerificationCode.objects.get(phone=phone)
    except VerificationCode.DoesNotExist:
        return False

    if row.expires_at < timezone.now():
        row.delete()
        return False

    if row.attempts >= 5:
        return False

    row.attempts += 1
    row.save(update_fields=["attempts"])

    if not secrets.compare_digest(row.code, code):
        return False

    row.delete()
    return True
